pfio/v2/pathlib.py

```diff
@@ -1,271 +1,756 @@
+    else:
+        return True
+
+
+def _removeprefix(text: str, prefix: str) -> str:
+    # NOTE: `str.removeprefix` supports version 3.9 or higher.
+    if text.startswith(prefix):
+        return text[len(prefix):]
+    return text
+
+
+def _compare_fs(lhs: FS, rhs: FS) -> bool:
+    if type(lhs) is type(rhs):
+        assert isinstance(lhs.cwd, str)
+        assert isinstance(rhs.cwd, str)
+
+        if isinstance(lhs, Local) and isinstance(rhs, Local):
+            return lhs.cwd == rhs.cwd
+        elif isinstance(lhs, S3) and isinstance(rhs, S3):
+            return (
+                lhs.cwd == rhs.cwd
+                and lhs.bucket == rhs.bucket
+                and lhs.endpoint == rhs.endpoint
+            )
+        elif isinstance(lhs, Hdfs) and isinstance(rhs, Hdfs):
+            return lhs.cwd == rhs.cwd and lhs.username == rhs.username
+        elif isinstance(lhs, Zip) and isinstance(rhs, Zip):
+            return lhs.cwd == rhs.cwd
+        else:
+            raise ValueError(f"unsupported FS: {lhs} and {rhs}")
 
-        self._parts = parts
+    return False
 
-    @property
-    def name(self):
-        if not self._parts:
-            return ""
-        if len(self._parts) == 1 and self._parts[0] == self.sep:
-            return ""
-        return os.path.normpath(self._parts[-1])
 
-    @property
-    def suffix(self):
-        return os.path.splitext(self.resolve())[1]
+def _not_supported(name: Optional[str] = None) -> NotImplementedError:
+    if not name:
+        import inspect
 
-    def with_suffix(self, ext):
-        assert ext.startswith('.')
-        name = self.name
-        if name == "":
-            raise ValueError(f"{self} has an empty name")
-        base, _ = os.path.splitext(name)
-        return self.parent / f"{base}{ext}"
+        stack = inspect.stack()
+        cname = stack[1].frame.f_locals["self"].__class__.__name__
+        fname = stack[1].function
+        name = f"{cname}.{fname}"
 
-    @property
-    def parent(self):
-        if not self._parts:
-            return self
+    return NotImplementedError(f"`{name}` is unsupported on this system")
 
-        if self._parts[0] == self.sep and len(self._parts) == 1:
-            return self
 
-        return Path(*(self._parts[:-1]), fs=self._fs)
+class PurePath(PathLike):
+    """
+    `pathlib.PurePosixPath` compatible interface.
+
+    Args:
+        args: construct paths.
+        fs: target file system.
+        scheme: specify URL scheme. (for `as_uri` method)
+
+    Note:
+        It conforms to `pathlib.PurePosixPath` of Python 3.12 specification.
+
+        this class not inherits any `pathlib` classes because
+        pfio filesystems is not suitable for pathlib abstact
+        and helper classes.
+
+    TODO:
+        `scheme` should moves to `FS`.
+    """
+
+    def __init__(
+        self,
+        *args: Union[str, PathLike],
+        fs: FS,
+        scheme: Optional[str] = None,
+    ) -> None:
+        if isinstance(fs, Local):
```

Comment:
Q: do you @yhrkw  have a plan to support [custom scheme](https://pfio.readthedocs.io/en/latest/design.html#custom-scheme) in near future? Currently, `pfio.v2.FS` does not have `.scheme` property, but it'd be worth trying adding the feature to FS class and using it here.

Reply:
Yes, I'm currently working on it. I will send a PR after merge.